### PR TITLE
Allow reference plugins to use > 2 wire operations and expvals

### PR DESCRIPTION
**Pull request by josh146**
**Description of the Change:**

* Removes limit on `default.qubit` which previously only allowed 1 and 2 qubit operations/expvals. Now, you can apply arbitrary N-qubit operations, and measure arbitrary N-qubit observables. This is done by replacing the `DefaultQubit.expand_one` and `DefaultQubit.expand_two` with a single, more general `expand` method, that works for any unitary/Hermitian matrix.

* Removes limit on `default.gaussian` which previously only allowed 1 and 2 mode gates. It now supports N-mode operations. This is done by replacing the `DefaultGaussian.expand_one` and `DefaultGaussian.expand_two` with a single, more general `expand` method, that works for any sized symplectic matrix.

**Benefits:**

* Allows `qml.Interferometer()` to work on the `default.gaussian` plugin for >2 modes
* Allows `qml.QubitUnitary()` to work on the `default.qubit` plugin for >2 modes
* Allows `qml.expval.Hermitian()` to work on the `default.qubit` plugin for >2 modes

**Possible Drawbacks:**

n/a

**Related GitHub Issues:**

#175

### CODE REVIEW

File: pennylane/plugins/default_gaussian.py [hunk 1]
```diff
@@ -765,13 +765,6 @@ def apply(self, operation, wires, par):
             self._state = self._operation_map[operation](*par, hbar=self.hbar)
             return # we are done here
 
-        if operation == 'Interferometer':
-            if par[0].shape[0] != len(wires):
-                raise ValueError("Interferomer unitary matrix applied to the incorrect "
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]: Isn't this still a useful error to raise?

File: pennylane/plugins/default_gaussian.py [hunk 2]
```diff
@@ -783,10 +776,7 @@ def apply(self, operation, wires, par):
         S = self._operation_map[operation](*par)
 
         # expand the symplectic to act on the proper subsystem
-        if len(wires) == 1:
-            S = self.expand_one(S, wires[0])
-        elif len(wires) == 2:
-            S = self.expand_two(S, wires)
+        S = self.expand(S, wires)
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]: nice! so much cleaner

File: pennylane/plugins/default_gaussian.py [hunk 3]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
-    def expand_one(self, S, wire):
-        r"""Expands a one-mode Symplectic matrix S to act on the entire subsystem.
+    def expand(self, S, wires):
+        r"""Expands a Symplectic matrix S to act on the entire subsystem.
 
         Args:
-            S (array): :math:`2\times 2` Symplectic matrix
-            wire (int): the wire S acts on
+            S (array): a :math:`2M\times 2M` Symplectic matrix
+            wires (Sequence[int]): the wires of modes S acts on
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
            wires (Sequence[int]): the wires of the modes that S acts on
```

File: pennylane/plugins/default_gaussian.py [hunk 4]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
 
-        ind = np.concatenate([np.array([wire]), np.array([wire])+self.num_wires])
-        rows = ind.reshape(-1, 1)
-        cols = ind.reshape(1, -1)
-        S2[rows, cols] = S.copy()
+        N = self.num_wires
+        w = np.asarray(wires)
 
-        return S2
+        if np.any(w < 0) or np.any(w >= N) or len(set(w)) != len(w):
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]: would the user be able to sneak non-`int` wires values in here?

File: pennylane/plugins/default_gaussian.py [hunk 5]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
+            raise ValueError('Bad target subsystems.')
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
            raise ValueError("Invalid target subsystems provided in 'wires' argument.")
```

File: pennylane/plugins/default_gaussian.py [hunk 6]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
-        S2 = np.identity(2*self.num_wires)
-        w = np.array(wires)
+        if M != len(wires):
+            raise ValueError('Incorrect number of subsystems for provided operation.')
 
-        S2[w.reshape(-1, 1), w.reshape(1, -1)] = S[:2, :2].copy() #X
-        S2[(w+self.num_wires).reshape(-1, 1), (w+self.num_wires).reshape(1, -1)] = S[2:, 2:].copy() #P
-        S2[w.reshape(-1, 1), (w+self.num_wires).reshape(1, -1)] = S[:2, 2:].copy() #XP
-        S2[(w+self.num_wires).reshape(-1, 1), w.reshape(1, -1)] = S[2:, :2].copy() #PX
+        S2[w.reshape(-1, 1), w.reshape(1, -1)] = S[:M, :M].copy()  # X
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
        S2[w.reshape(-1, 1), w.reshape(1, -1)] = S[:M, :M].copy()  # XX
```

File: pennylane/plugins/default_gaussian.py [hunk 7]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
+        S2[(w + N).reshape(-1, 1), (w + N).reshape(1, -1)] = S[M:, M:].copy()  # P
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
        S2[(w + N).reshape(-1, 1), (w + N).reshape(1, -1)] = S[M:, M:].copy()  # PP
```

File: pennylane/plugins/default_qubit.py [hunk 8]
```diff
@@ -388,78 +376,44 @@ def reset(self):
...
-        perm = np.r_[p, p+3]
-        # reshape U into another array which has one index per subsystem, permute dimensions, back into original-shape array
-        temp = np.prod(dim)
-        U = U.reshape(dim * 2).transpose(perm).reshape([temp, temp])
-        U = np.kron(np.kron(np.eye(before), U), np.eye(after))
-        return U
-
-    def expand_multi(self, U, wires):
-        r"""Expand a multi-qubit operator into a full system operator.
+        # generate N qubit basis states via the cartesian product
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]: use `itertools`, it'll likely be fastest

File: pennylane/plugins/default_qubit.py [hunk 9]
```diff
@@ -388,78 +376,44 @@ def reset(self):
...
+        tuples = np.stack(np.meshgrid(*[[0, 1] for _ in range(N)]), -1).reshape(-1, N)
+        tuples[:, [0, 1]] = tuples[:, [1, 0]]
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]: what is this line doing?

File: pennylane/plugins/default_qubit.py [hunk 10]
```diff
@@ -388,78 +376,44 @@ def reset(self):
...
-        """
-        raise NotImplementedError
+        # expand U to act on the entire system
+        U = np.kron(U, np.identity(2**len(inactive_wires)))
+
+        # move active wires to beginning of the list of wires
+        rearanged_wires = np.array(list(wires)+inactive_wires)
+
+        # convert to computational basis
+        perm = np.sum(2**np.arange(N-1, -1, -1)*tuples[:, rearanged_wires], axis=1)
```
Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]: using `np.cumprod` might be more efficient?

Comment (co9olguy) [inline pennylane/plugins/default_qubit.py]: Hard to parse what this is doing. I feel like numpy index (un)raveling might make it more clear what is happening?

File: tests/test_default_gaussian.py [hunk 11]
```diff
@@ -394,14 +394,14 @@ def test_apply_errors(self):
             p = [thermal_state(0.5)]
             self.dev.apply('GaussianState', wires=[0], par=[p])
 
-        with self.assertRaisesRegex(ValueError, 'incorrect number of subsystems'):
+        with self.assertRaisesRegex(ValueError, 'Incorrect number of subsystems'):
             p = U
             self.dev.apply('Interferometer', wires=[0], par=[p])
 
-        with self.assertRaisesRegex(ValueError, 'Only 2-mode interferometers are currently supported.'):
+        with self.assertRaisesRegex(ValueError, 'Bad target subsystems'):
```
Comment (co9olguy) [inline tests/test_default_gaussian.py]: If you take suggestion for new error message above, will need to update this assert

File: tests/test_default_qubit.py [hunk 12]
```diff
@@ -264,71 +266,90 @@ def test_expand_one(self):
...
-        # test exception raised if U is not 4x4 matrix
-        with self.assertRaisesRegex(ValueError, "4x4 matrix required"):
-            dev.expand_two(U, [0])
-
-        # test exception raised if two subsystems are not provided
-        with self.assertRaisesRegex(ValueError, "Two target subsystems required"):
-            dev.expand_two(U2, [0])
-
         # test exception raised if unphysical subsystems provided
         with self.assertRaisesRegex(ValueError, "Bad target subsystems."):
```
Comment (co9olguy) [inline tests/test_default_qubit.py]: same point about changing error message

File: tests/test_default_qubit.py [hunk 13]
```diff
@@ -264,71 +266,90 @@ def test_expand_one(self):
...
+        rows = np.array([0, 4, 1, 5, 2, 6, 3, 7])
+        expected = np.kron(I, U_toffoli[:, rows][rows])
+        self.assertAllAlmostEqual(res, expected, delta=self.tol)
+
+        # test applied to wire 3, 0, 2
+        res = dev.expand(U_toffoli, [3, 0, 2])
+        # change the control qubit on the Toffoli gate
+        rows = np.array([0, 4, 1, 5, 2, 6, 3, 7])
+        expected = np.kron(U_swap, np.kron(I, I)) @ np.kron(I, U_toffoli[:, rows][rows]) @ np.kron(U_swap, np.kron(I, I))
+        self.assertAllAlmostEqual(res, expected, delta=self.tol)
```
Comment (co9olguy) [inline tests/test_default_qubit.py]: except for `expected`, feels like these could be combined into one test which is run with multiple variants

File: pennylane/plugins/default_gaussian.py (hunk 4 above)
Comment (josh146) [reply pennylane/plugins/default_gaussian.py]: I don't think so, there is validation in the `Operation` class to catch this.

File: pennylane/plugins/default_qubit.py (hunk 9 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Using this NumPy trick to calculate the cartesian product results in the first two columns being in the wrong order... perhaps a reason to switch to itertools

File: tests/test_default_gaussian.py (hunk 11 above)
Comment (josh146) [reply tests/test_default_gaussian.py]:
```suggestion
        with self.assertRaisesRegex(ValueError, "Invalid target subsystems provided in 'wires' argument"):
```

File: tests/test_default_qubit.py (hunk 12 above)
Comment (josh146) [reply tests/test_default_qubit.py]:
```suggestion
        with self.assertRaisesRegex(ValueError, "Invalid target subsystems provided in 'wires' argument"):
```

File: tests/test_default_qubit.py (hunk 13 above)
Comment (josh146) [reply tests/test_default_qubit.py]: Yep I agree 👍 
This test hasn't been ported to pytest yet. Once that is done, it will be easier (we can just parametrize the test data/expected results, and pass it via a fixture).

File: pennylane/plugins/default_qubit.py (hunk 10 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Essentially, I am just converting the basis state bit string into a decimal number. I.e., `[0, 1, 0, 1, 1] = 2^3+2^1+2^0 = 11`.

Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: perfect use-case for ravel/unravel

File: pennylane/plugins/default_gaussian.py [hunk 14]
```diff
@@ -795,42 +785,36 @@ def apply(self, operation, wires, par):
...
+        S2[w.reshape(-1, 1), (w + N).reshape(1, -1)] = S[:M, M:].copy()  # XP
+        S2[(w + N).reshape(-1, 1), w.reshape(1, -1)] = S[M:, :M].copy()  # PX
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/plugins/default_gaussian.py]: I didn't expect this to work by simply replacing the '2' with 'M', was in shock when the tests passed 😆

File: pennylane/plugins/default_qubit.py (hunk 8 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Yep!
```python
In [17]: %timeit np.array(list(product([0, 1], repeat=4)))
9.87 µs ± 347 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [18]: %timeit np.stack(np.meshgrid(*[[0, 1] for _ in range(4)]), -1).reshape(-1, 4)
62.5 µs ± 1.38 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

File: pennylane/plugins/default_qubit.py (hunk 9 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Replaced these two lines with 
```python
# generate N qubit basis states via the cartesian product
tuples = np.array(list(itertools.product([0, 1], repeat=N)))
```

File: pennylane/plugins/default_gaussian.py (hunk 1 above)
Comment (josh146) [reply pennylane/plugins/default_gaussian.py]: It's been superseded by the more general one on line 812, that works for any operation (and doesn't need to single out the interferometer)

File: pennylane/plugins/default_qubit.py (hunk 10 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: I'm not sure I fully follow how it would work with ravel/unravel?

File: pennylane/plugins/default_qubit.py (hunk 8 above)
Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: lol also much clearer what it's doing :)

File: pennylane/plugins/default_qubit.py (hunk 10 above)
Comment (josh146) [reply pennylane/plugins/default_qubit.py]: I did a Google search, and found an alternative implementation that is slightly faster:
```python
In [47]: %timeit np.sum(2**np.arange(N-1, -1, -1)*tuples[:, rearanged_wires], axis=1)
12.6 µs ± 280 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [48]: %timeit np.right_shift(np.packbits(tuples[:, rearanged_wires], -1), tuples[:, rearanged_wires].size).squeeze()//2**N
7.49 µs ± 133 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

But I feel that the speeds are so similar, that it isn't much point changing.

Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: Sorry, it's `ravel_multi_index` that you want
```np.ravel_multi_index([0,1,0,1,1],[2]*5)```

Comment (josh146) [reply pennylane/plugins/default_qubit.py]: Oh that is awesome!

Comment (josh146) [reply pennylane/plugins/default_qubit.py]:
```python
In [61]: %timeit np.ravel_multi_index(tuples[:, rearanged_wires].T, [2]*4)
3.46 µs ± 59 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
Also the fastest implementation!

Comment (co9olguy) [reply pennylane/plugins/default_qubit.py]: yeah the numpy docs for this function do it a disservice. Even I, who knew what I was looking for, still had to scratch my head a bit to implement it

**Comment by josh146** [comment]
All suggested changes made.
